source/extensions/filters/http/aws_lambda/api_gateway_transformer.cc

```diff
@@ -0,0 +1,164 @@
+    return;
+  }
+
+  // downcast RequestResponseHeaderMap to ResponseHeaderMap
+  Http::ResponseHeaderMap* response_headers = static_cast<Http::ResponseHeaderMap*>(&header_map);
+
+  // all information about the request format is to be contained in the response body
+  
+  // parse response body as JSON
+  auto bodystring = body.toString();
+  nlohmann::json json_body;
+  try {
+    json_body = json::parse(bodystring);
+  } catch (json::exception exception){
+    ApiGatewayError error = {400, "400", "failed to parse response body as JSON"};
+    return ApiGatewayTransformer::format_error(*response_headers, body, error);
+  }
+
+  // set response status code
+  if (json_body.contains("statusCode")) {
+    uint64_t status_value;
+    try {
+      status_value = json_body["statusCode"].get<uint64_t>();
+    } catch (json::exception exception){
+      ApiGatewayError error = {400, "400", "Non-integer status code"};
+      return ApiGatewayTransformer::format_error(*response_headers, body, error);
+    }
+    response_headers->setStatus(status_value);
+  } else {
+    response_headers->setStatus(DEFAULT_STATUS_VALUE);
+  }
+
+  // set response headers
+  if (json_body.contains("headers")) {
+    auto headers = json_body["headers"];
+    if (!headers.is_object()) {
+        ENVOY_LOG(debug, "invalid headers object");
+        ApiGatewayError error = {400, "400", "invalid headers object"};
+        return ApiGatewayTransformer::format_error(*response_headers, body, error);
+    }
+    for (json::iterator it = headers.begin(); it != headers.end(); it++) {
+        auto header_key = it.key();
+        auto header_value = it.value().get<std::string>();
+        add_response_header(*response_headers, header_key, header_value, false);
+    }
+  }
+
+  // set multi-value response headers
+  if (json_body.contains("multiValueHeaders")) {
+    auto multi_value_headers = json_body["multiValueHeaders"];
+    if (!multi_value_headers.is_object()) {
+        ENVOY_LOG(debug, "invalid multi headers object");
+        ApiGatewayError error = {400, "400", "invalid multi headers object"};
+        return ApiGatewayTransformer::format_error(*response_headers, body, error);
+    }
+
+    for (json::iterator it = multi_value_headers.begin(); it != multi_value_headers.end(); it++) {
+        auto header_key = it.key();
+        auto lower_case_header_key = Envoy::Http::LowerCaseString(header_key);
+        auto header_values = it.value();
+
+        for (json::iterator inner_it = header_values.begin(); inner_it != header_values.end(); inner_it++) {
+          auto header_value = inner_it.value().get<std::string>();
+          add_response_header(*response_headers, header_key, header_value, true);
+        }
+    }
+  }
+
+  // set response body
+  body.drain(body.length());
+  if (json_body.contains("body")) {
+    std::string body_dump;
+    try {
+      body_dump = json_body["body"];
+    } catch (json::exception exception){
+      body_dump = json_body["body"].dump();
```

Comment:
why do we need to call dump here? is there to find that out without exceptions?

Reply:
Basically this is a test to determine if the body is a string (in which case `json_body["body"]` can be assigned to a string value without issue) or if it's a JSON object (in which case we need to dump the body to a string). I think I can use `json_body["body"].is_object()` to accomplish this

Reply:
why not always use dump()?

Reply:
is it because dump will add quotes/escape to a string?